janus_core/md.py

```diff
@@ -471,18 +474,25 @@ def _write_traj(self) -> None:
                 append=append,
             )
 
-    def _write_restart(self) -> None:
+    def _write_restart(self, final: bool = False) -> None:
         """Write restart file and (optionally) rotate files saved."""
         step = self.offset + self.dyn.nsteps
         if step > 0:
-            restart_file = f"{self.restart_stem}-{step}.xyz"
+            if not self.restart_stem:
+                if not final:
+                    restart_file = f"{self.file_prefix}-T{self.temp}-res-{step}.xyz" 
+                else:
+                    restart_file = f"{self.file_prefix}-T{self.temp}-final.xyz"
+            else:
+                restart_file = f"{self.restart_stem}-res-{step}.xyz"
+            columns = ["symbols", "positions", "momenta", "masses"] if not final else None
```

Comment:
Why do we want `final` and `not final` to be save different things?

Reply:
mainly because one is restart the other one not a restart, the issue of extra info printed is an ase one I will not be too bothered about...

Reply:
If the `final` files are saved in order to store configurations consistent with that different temperatures, does that not need more or less the same information as for restarting?

I suppose we don't need the current momenta, but it's currently stored anyway (in fact, everything in the restart is included, as well as the extra `spacegroup_kinds`), so I'm not sure I understand what we gain from distinguishing the two?

Reply:
It's still not completely clear to me exactly how specifically we would expect the two to differ. I can see in principle we may want to write out some things at the end rather than many times, but since the information essentially needs to be all be there in order to restart a simulation, I'm not sure how much there is to gain.

(Arguably just removing writing out `spacegroup_kinds` is not insignificant, but then it's unclear it there's any benefit in ever writing it out?)

I was actually just thinking about your second point in terms of including these changes within `_write_restart`. If the `final` outputs serve a different purpose, are formatted differently, and have no need to be worried about at a regular interval or influenced by any of the `restart` options, I think it would make a lot more sense to define it elsewhere in the code.

If there's any significant amount of code duplication, they can always call a separate shared function.

Reply:
by having a final we address two issues.
1. issue in the heating when you want to have the last configuration for each T ramp to be reused as input file for another run. there is the issue that restart can coincide with the final. that is a duplication we may want to avoid or we may not want to avoid. if one wants to avoid it will be a conflict with the restart in terms of meanings.
2. we address the point when restart frequencies, does not include the last step.
in terms of information restart and final may contain different information... final is to start a new simulation and not to restart it. restart will contain more info than a final.. so one can use restart for the purpose of final, given the steps align correctly.

I am happy to have them implemented separately but since this somehow addresses a mistmatch in restart I am not opposed in addressing it there, @harveydevereux what are your thoughts?

Reply:
I do think having some method ```_write_final``` or another name to handle the case of final is better in my opinion, even if the data is the same, because we need to invoke it in special circumstances.

The crux as Elliott is pointing out is what (if any) additional data needs to be in a ```-final``` over a  ```-res```, and why? I'm not familiar enough with Janus yet to make the decision.

I don't actually know what ```columns``` does, its not documented in ASE write, I can't find it anywhere in ASE IO actually. 

```
            write(
                restart_file,
                self.struct,
                write_info=True,
                columns=columns
            )
```

Reply:
Currently working on this and also addressing some things Jacob mentioned.

I am going to assume from now that final should take the same columns, in pseudo-code

```python
write(
    final_file, 
    self.struct,  
    write_info=True, 
    columns=["symbols", "positions", "momenta", "masses"]
)
```

Unless there is a specific reason posted on this thread to not do this, there are some

we can always do it in future

Reply:
No objections here.

It may well be something to give users more control of at some point too.

Reply:
done this in latest push